src/doc_iter.rs

```diff
@@ -0,0 +1,309 @@
+use crate::document::Document;
+
+impl Document {
+    /// Returns an iterator over all all document nodes including
+    /// comments and fragments.
+    ///
+    /// When encountering a container node (mapping, sequence or fragment),
+    /// the container node is yielded first, then all of its children.
+    pub fn iter(&self) -> DocIter {
+        let v = std::slice::from_ref(self);
+        DocIter {
+            stack: vec![v.iter()],
+        }
+    }
+
+    /// Returns an iterator over all value nodes in the document.
+    /// The iterator yields tuples of (object-path, value-node).
+    pub fn iter_path(&self) -> DocPathIter {
+        let v = std::slice::from_ref(self);
+        DocPathIter {
+            stack: vec![v.iter()],
+            aggregate: Vec::new(),
+            path: Vec::new(),
+        }
+    }
+
+    /// Returns a mutable iterator over all value nodes in the document.
+    /// The iterator yields tuples of (object-path, value-node).
+    pub fn iter_path_mut(&mut self) -> DocPathIterMut {
+        let v = std::slice::from_mut(self);
+        DocPathIterMut {
+            stack: vec![v.iter_mut()],
+            aggregate: Vec::new(),
+            path: Vec::new(),
+        }
+    }
+}
+
+impl<'a> IntoIterator for &'a Document {
+    type Item = &'a Document;
+    type IntoIter = DocIter<'a>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.iter()
+    }
+}
+
+pub struct DocIter<'a> {
+    stack: Vec<std::slice::Iter<'a, Document>>,
+}
+
+impl<'a> Iterator for DocIter<'a> {
+    type Item = &'a Document;
+    fn next(&mut self) -> Option<Self::Item> {
```

Comment:
`next()` is pretty deeply nested, and it's hard to see which braces at the end of the function correspond to each control statement. It looks like its logic can be broken into two sequential steps:

1. Find the first document pointed to by `stack` (with an early return if `stack` is emptied).
2. `match` on that document.

In code:

```rust
    fn next(&mut self) -> Option<Self::Item> {
        // This loop retrieves the first document in self.stack.
        let val = loop {
            let top = match self.stack.last_mut() {
                Some(top) => top,
                None => return None,  // self.stack is empty
            };
            if let Some(val) = top.next() {
                break val  // We found the first document self.stack points to
            } else {
                // self.stack's top element is empty; remove it and try the
                // next.
                self.stack.pop();
                continue;
            }
        };
        match val {
            // omitted
        }
```

Reply:
Oh, and I think this technique applies to DocPathIter and DocPathIterMut too.

Reply:
Thanks, `DocIter` is much shorter using the "loop-breaks-with-value" construction.

I'm not sure that `DocPathIter` can use the same technique because unlike `DocIter` which gives you every node, `DocPathIter` wants to hide some of the nodes in favor of returning a `(path, &Document)` tuple.  The "useless" nodes like `Comment` and `Compact` are not returned and the `Fragment` node is transformed into either (list-index, value) or (mapping-key, value).

Reply:
BTW, I'd really like the type of `DocPathIter{,Mut}::Item` to be:
```
type Item = (&[DocPath<'path>], &'a Document)
```
But I don't know how to write the lifetime parameter for `DocPath` or `&[DocPath]`.  I think I want "self" lifetime because those values live as long as the existence of the iterator (where-as the `Document` references live as long as the document).

Any ideas?

Reply:
I don't think that is possible with the `Iterator` trait. See https://rust-lang.github.io/rfcs/1598-generic_associated_types.html for more details (I think the trait you want is the `StreamingIterator` trait from the Motivation heading).